### PR TITLE
chip_data_model.gni: add option for custom-implemented clusters

### PR DESCRIPTION
**Pull request by plan44**
Add `zap_clusters_with_custom_implementation` option to `chip_data_model()` template, which allows to list clusters that need to be available in a device, but don't use the standard implementation in `src/app/clusters/<clustername>/<clustername>.cpp`.

# Usage
To exclude a particular cluster's standard implementation in favor of an application level implementation:
- in the BUILD.gn file for the application specific `.zap` file
- in the instantiation of the `chip_data_model` template
- add a variable `zap_clusters_with_custom_implementation`
- assign it a list of strings with cluster names, corresponding to directories in the `src/app/clusters` directory.
- this causes the cluster implementation file(s) not to be included in the build, allowing re-implementation at the application level.

Example, excluding standard implementation of level-control:
```
chip_data_model("zap") {
  zap_file = "myproject.zap"
  zap_clusters_with_custom_implementation = [ "level-control" ]
  zap_pregenerated_dir = "//zap-generated"
}
```

# Background
Some clusters, for example Level Control, are implemented with the assumption of direct low-level access to the device hardware. For bridged devices, the actual interface might be much more high level, as in my case where I have lights which are fully capable of performing parametrized smooth transitions but are NOT capable of receiving output changes in millisecond intervals. In such cases, a cluster might need application-specific re-implementation.

This changeset allows, from the application's ZAP BUILD.gn, to exclude the standard cluster implementation file, allowing re-implementation of the needed ember callbacks in a application specific file outside the connectedhomeip repo.

#### Testing
* tested to not affect normal builds not using this feature
* tested with an application which re-implements level-control in a separate .cpp file. Without the feature, linking has duplicate symbol errors (from both custom and standard implementations) - using the feature as shown in the example above, the app links and runs ok.

### CODE REVIEW

**Review by andy31415 — DISMISSED** [review]
Marking changes requested to wait until after 1.0 branch

**Comment by plan44** [comment]
I think this is an important feature. Its further review was postponed to post-1.0, which would be now ;-)

**Comment by andy31415** [comment]
> I still think this is an important feature. Its further review was postponed to post-1.0, which would be now ;-)

The PR is marked as draft - should this not be draft anymore?
It probably also suffers from being very far down the open PR list ... less likely to get reviews.

**Comment by andy31415** [comment]
@plan44 - please remove this from draft if that is appropriate - that should get the bot start asking for reviews.

File: src/app/chip_data_model.gni [hunk 1]
```diff
@@ -99,8 +100,23 @@ template("chip_data_model") {
                                      [ invoker.zap_file ])
     }
 
+    _cust_impl_clusters = []
+    if (defined(invoker.zap_clusters_with_custom_implementation)) {
+      _cust_impl_clusters =  invoker.zap_clusters_with_custom_implementation
```
[the remark this answers is not in the document]
Comment (plan44) [reply src/app/chip_data_model.gni:105]: Yes, this works as is, without having `zap_clusters_with_custom_implementation` (tests) and also when actually using the feature (my bridge, having custom level-control implementation).

I'm a total gn newbie and if I remember correctly I gleaned how to implement that from somewhere else. Maybe these local vars don't have the re-assignment restriction?

**Comment by plan44** [comment]
> @plan44 - please remove this from draft if that is appropriate - that should get the bot start asking for reviews.

I will, as soon as I find time to rebase and re-test on master. Thanks for having had a look already :-)

File: src/app/chip_data_model.gni (hunk 1 above)
[the remark this answers is not in the document]
Comment (plan44) [reply src/app/chip_data_model.gni]: I just rebased the PR onto master, re-ran test locally (check target) successfully without gn complaining, so I removed the draft status now. Hope this is ok.

**Comment by plan44** [comment]
@andy31415 I've rebased and retested a few days ago, and removed the draft status. Checks look fine except for pullaprove/license/cla, which I do not understand, as I did agree to the CLA a while ago and some PRs of mine are merged already. I have no idea how to re-agree to the CLA, if that's what the status is asking for.

**Comment by bzbarsky-apple** [comment]
@plan44, @andy31415 is on vacation right now, as is @woody-apple, and they're the obvious people to be able to override the CLA bits....

In practice you probably want to ping people after Jan 3.

**Comment by plan44** [comment]
@bzbarsky-apple thanks for the info - no problem waiting until next year. Or rather than waiting, maybe finding time to prepare the libev and gaps-in-endpointId PRs ;-)

File: src/app/chip_data_model.gni [hunk 2]
```diff
@@ -137,8 +138,22 @@ template("chip_data_model") {
                                      [ invoker.zap_file ])
     }
 
+    _cust_impl_clusters = []
```
Comment (bzbarsky-apple) [inline src/app/chip_data_model.gni:141]:
```suggestion
    _custom_impl_clusters = []
```

File: src/app/chip_data_model.gni [hunk 3]
```diff
@@ -137,8 +138,22 @@ template("chip_data_model") {
...
+    if (defined(invoker.zap_clusters_with_custom_implementation)) {
+      _cust_impl_clusters = invoker.zap_clusters_with_custom_implementation
+    }
+
     foreach(cluster, _cluster_sources) {
-      if (cluster == "door-lock-server") {
+      _cust_impl = false
```
Comment (bzbarsky-apple) [inline src/app/chip_data_model.gni:147]:
```suggestion
      _custom_impl = false
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/chip_data_model.gni (hunk 2 above)
Comment (plan44) [reply src/app/chip_data_model.gni]: agreed, `cust` is too much abbreviated, `custom` is better.
I just changed all occurrences and updated the PR

**Review by jmartinez-silabs — APPROVED** [review]

**Pull request merged**